sonic-sairedis : Wred stats feature changes on Sai-redis and Syncd

rpmarvell opened the pull request:
* Stats capability query API support is added

Details : 
- Changes are done to support the stats capability query from Swss to SAI.
- This was implemented as part of WRED and ECN statistics
- Statistics capability query is implemented in sairedis/syncd

HLD : https://github.com/sonic-net/SONiC/blob/master/doc/qos/ECN_and_WRED_statistics_HLD.md

Expected order of pull-request commits :

1) sonic-swss common pull request : https://github.com/sonic-net/sonic-swss-common/pull/777
2) sonic-yang-model pull requests : https://github.com/sonic-net/sonic-buildimage/pull/14758
3) sonic-sairedis pull request : https://github.com/sonic-net/sonic-sairedis/pull/1234
4) sonic-swss : pull request : https://github.com/sonic-net/sonic-swss/pull/2750
5) sonic-utilities pull request : https://github.com/sonic-net/sonic-utilities/pull/2807

kcudnik commented:
please fix errors

rpmarvell commented:
> 
This PR is dependent on swss-common PR 777 as i have mentioned in the description. Hence the errors. Could you please review and approve the swss-common PR if possible..! Thank you

stephenxs commented on lib/ClientSai.cpp:
```diff
@@ -999,6 +989,149 @@ sai_status_t ClientSai::waitForGetStatsResponse(
+        {
+		/* Populate stat_enum */
+            size_t old_stat_enum_position = stat_enum_position;
+            stat_enum_position = stat_enum_str.find(",", old_stat_enum_position);
```
Is it possible to leverage `tokenize` which returns all the substring split by a splitter, eg `,`?

rpmarvell replied:
Tokenize should also work fine. I have opted this, as this algorithm is used in various Attribute/enum processing APIs (including waitForQueryAattributeEnumValuesCapabilityResponse). I would stick to current way of processing the Attributes/Enums.

stephenxs replied:
Hi @rpmarvell 
could you please clarify what optimization in your implementation compared to using `tokenize`?
In case the current way is better, I would like to suggest implementing a common function (maybe in the base class) and calling it in `ClientSai`, `RedisRemoteSaiInterface` instead of duplicating it in two sub classes.

stephenxs commented on lib/RedisRemoteSaiInterface.cpp:
```diff
@@ -1243,6 +1233,151 @@ sai_status_t RedisRemoteSaiInterface::waitForGetStatsResponse(
+        {
+		/* Populate stat_enum */
+            size_t old_stat_enum_position = stat_enum_position;
+            stat_enum_position = stat_enum_str.find(",", old_stat_enum_position);
```
Suggest using `tokenize`

rpmarvell replied:
Tokenize should also work fine. I have opted this, as this algorithm is used in various Attribute/enum processing APIs (including waitForQueryAattributeEnumValuesCapabilityResponse). I would stick to current way of processing the Attributes/Enums.

rpmarvell commented on lib/ClientSai.cpp:
```diff
@@ -999,6 +989,149 @@ sai_status_t ClientSai::waitForGetStatsResponse(
+        size_t stat_modes_position = 0;
+        for (uint32_t i = 0; i < num_capabilities; i++)
+        {
+		/* Populate stat_enum */
```
I have fixed it.

rpmarvell commented:
> please fix build errors

sonic-sairedis is dependent on sonic-swss-common PR [https://github.com/sonic-net/sonic-swss-common/pull/777](https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sonic-2Dnet_sonic-2Dswss-2Dcommon_pull_777&d=DwMFAg&c=nKjWec2b6R0mOyPaz7xtfQ&r=hoD2LbGQ9KQ1NPvrVvsseLUTyo_X-OJH7_qVkrYUdvw&m=vBjBqZWfL6Oqlgrlg08-CEirIIyF2ONv4WPQy_AvH9YLK-FI2SjxfTJICYAF7Mky&s=VqD7fYDxu8_zKlW5A9EtQq9a2AkwbifbwgkC9GCkujQ&e=) .
This sonic-swss-common PR has already been approved by maintainer but we are facing issues in merging this PR as the CI is unstable. 
There is an issue raised for this in the SONiC community by Myron Sosyak, Link for this issue : https://github.com/sonic-net/sonic-swss-common/issues/784

stephenxs commented:
Hi @rpmarvell 
Please refer to my [latest comment](https://github.com/sonic-net/sonic-sairedis/pull/1234#discussion_r1209374839). I'm not able to unresolve a comment
Thanks

akokhan commented:
/azp run

akokhan commented:
/azp run

akokhan commented:
@rpmarvell , even swss-common changes are merged, sairedis CI is still failing. Please check.

rpmarvell commented:
> Hi @rpmarvell Please refer to my [latest comment](https://github.com/sonic-net/sonic-sairedis/pull/1234#discussion_r1209374839). I'm not able to unresolve a comment Thanks

I have addressed your review comment. Thank you!

stephenxs reviewed (commented):
LGTM. Please add more unit test cases to [cover the code change](https://dev.azure.com/mssonic/build/_build/results?buildId=287225&view=codecoverage-tab)

kcudnik commented:
please add some unittests to meet code coverage threshold

rpmarvell commented:
> please add some unittests to meet code coverage threshold
? 


> LGTM. Please add more unit test cases to [cover the code change](https://dev.azure.com/mssonic/build/_build/results?buildId=287225&view=codecoverage-tab)

I am not finding any infra to add unit tests for changes syncd/Syncd.cpp and lib/RedisRemoteSaiInterface.cpp 
As these functionalities lies in private functions of class, not even able to call them from UT. Any suggestions are appreciated

zhangyanzhao commented:
@stephenxs  @msosyak can you please help to review and approve this PR? Thanks.

zhangyanzhao commented:
@rpmarvell  can you please help to address the build failure? Thanks.

stephenxs commented:
Hi @rpmarvell 
I'm OK with the code.
Could you please add more test cases to satisfy the coverage?
```
Total: 533 lines
Missing: 386 lines
Coverage: 27%
Threshold: 80%
[Diff coverage](https://dev.azure.com/mssonic/be1b070f-be15-4154-aade-b1d3bfb17054/_build/results?buildId=297771&view=codecoverage-tab)
```

stephenxs commented:
Hi @rpmarvell 
Could you please update the DD when the UT will be ready?
Thanks

kcudnik commented:
please resolve conflicts

rpmarvell commented:
I have addressed the UT. Could you please help merge this change with required approvals.!

kcudnik commented:
please fix build errors

rpmarvell commented:
> please fix build errors
> 
> ```
> In file included from orchdaemon.h:51,
>                  from main.cpp:25:
> dash/dashrouteorch.h:27:11: error: 'route' in namespace 'dash' does not name a type
>    27 |     dash::route::Route metadata;
>       |           ^~~~~
> dash/dashrouteorch.h:46:11: error: 'route' in namespace 'dash' does not name a type
>    46 |     dash::route::Route metadata;
>       |           ^~~~~
> make[4]: *** [Makefile:918: orchagent-main.o] Error 1
> make[4]: *** Waiting for unfinished jobs....
> In file included from orchdaemon.h:51,
>                  from orchdaemon.cpp:5:
> dash/dashrouteorch.h:27:11: error: 'route' in namespace 'dash' does not name a type
>    27 |     dash::route::Route metadata;
>       |           ^~~~~
> dash/dashrouteorch.h:46:11: error: 'route' in namespace 'dash' does not name a type
>    46 |     dash::route::Route metadata;
>       |           ^~~~~
> ```
> 
> error is in swss, mybe not related to this change


This is an existing break in sonic-swss, not introduced by the changes in this PR.
Even the recent PR https://github.com/sonic-net/sonic-sairedis/pull/1329 is having this build break..!

rpmarvell commented:
@kcudnik , @stephenxs  : Could you please approve and merge the changes as all tests are passing. Thank you..!

kcudnik commented on lib/ClientSai.cpp:
```diff
@@ -999,6 +989,104 @@ sai_status_t ClientSai::waitForGetStatsResponse(
+
+    auto switchIdStr = sai_serialize_object_id(switchId);
+    auto objectTypeStr = sai_serialize_object_type(objectType);
+    const std::string list_size = std::to_string(stats_capability->count);
```
if stats_capability is null this will crash, you dont check for null, also follow standard use listSize name like other

kcudnik commented on lib/ClientSai.cpp:
```diff
@@ -999,6 +989,104 @@ sai_status_t ClientSai::waitForGetStatsResponse(
+
+    m_communicationChannel->set(switchIdStr, entry, REDIS_ASIC_STATE_COMMAND_STATS_CAPABILITY_QUERY);
+
+    auto status = waitForQueryStatsCapabilityResponse(stats_capability);
+
+    return status;
```
just return wait...

kcudnik commented on lib/Recorder.cpp:
```diff
@@ -1093,6 +1111,67 @@ void Recorder::recordQueryAattributeEnumValuesCapabilityResponse(
+                object_type_str.c_str(),
+                list_size.c_str());
+
+    recordQueryStatsCapability (key, values);
```
space not needed before (

kcudnik commented on lib/RedisRemoteSaiInterface.cpp:
```diff
@@ -1247,6 +1237,106 @@ sai_status_t RedisRemoteSaiInterface::waitForGetStatsResponse(
+
+    auto switchIdStr = sai_serialize_object_id(switchId);
+    auto objectTypeStr = sai_serialize_object_type(objectType);
+    const std::string list_size = std::to_string(stats_capability->count);
```
not checking stats_capability) for null

rpmarvell replied:
Addressed the comment

kcudnik commented on meta/Meta.cpp:
```diff
@@ -986,37 +994,6 @@ sai_status_t Meta::meta_validate_stats(
     return SAI_STATUS_SUCCESS;
 }
 
-sai_status_t Meta::meta_validate_query_stats_capability(
```
why this is removed ?

rpmarvell replied:
This had the same MACRO checks as we do in the actual functions. If it is mandatory to keep this function alive, i can add it back.

kcudnik commented on meta/SaiSerialize.cpp:
```diff
@@ -2626,6 +2626,66 @@ std::string sai_serialize_redis_communication_mode(
+	}
+    }
+
+    j["stat_modes"] = arr;
```
you are serializing modes as array, not sure if this should be in this way or just and integer value

kcudnik commented on meta/SaiSerialize.cpp:
```diff
@@ -5117,6 +5177,82 @@ void sai_deserialize_redis_communication_mode(
 }
 
+
+/*************************************************************
+* Function    : sai_deserialize_stats_capability_list
+* Parameters  :
+*              stats_capability - stats capability enum list
+*              stat_enum_str    - SAI stat enum list as string
+*              stat_modes_str   - SAI stat mode list as string
+* Description :
+*              1. Iterate thru stat_enum_str and populate
+*                 the stats_capability with respective
+*                 stat_enum (String to Enum conversion)
+*              2. Iterate thru stat_modes_str and populate
+*                 the stats_capability with respective
+*                 stat_modes (String to Enum conversion)
+**************************************************************/
```
if you want to add comments, add it in doxygen format as rest of the project

kcudnik reviewed (changes requested):
please address comments

stephenxs commented:
@rpmarvell  could you please address Kamil's comments? thanks.

zhangyanzhao commented:
@rpmarvell can you please address Kamil's comments? Those need be addressed before merging.

stephenxs commented:
@rpmarvell  kindly reminder to address Kamil's comments. Thanks

rpmarvell commented:
> @rpmarvell kindly reminder to address Kamil's comments. Thanks

I am working on addressing the comments. Will be updating this thread once the changes are ready.
Thank you.

kcudnik commented on lib/RedisRemoteSaiInterface.cpp:
```diff
@@ -1375,6 +1365,113 @@ sai_status_t RedisRemoteSaiInterface::waitForGetStatsResponse(
+        return SAI_STATUS_INVALID_PARAMETER;
+    }
+
+    if (stats_capability && stats_capability->list)
```
need to check if list is null and count non zero, this is invalid scenario

kcudnik replied:
also if stats_capability is NULL, then you skip if and in line 1387 you will have exception for page fault

rpmarvell replied:
I have added count non-zero check.  if stats_capability is NULL, this function will return from line 1381 only.

kcudnik commented on meta/SaiSerialize.cpp:
```diff
@@ -5440,3 +5500,79 @@ void sai_deserialize_redis_link_event_damping_aied_config(
     sai_deserialize_number(j["flap_penalty"], value.flap_penalty, false);
 }
+
+/** @brief deserializes the stats capability list 
```
brieff should be in next line

rpmarvell replied:
I have taken care of this in recent code changes.

kcudnik commented on meta/SaiSerialize.cpp:
```diff
@@ -5440,3 +5500,79 @@ void sai_deserialize_redis_link_event_damping_aied_config(
+*   @param stat_modes_str SAI stat mode list as string
+*   @return Void
+*/
+void sai_deserialize_stats_capability_list(
+        _Inout_ sai_stat_capability_list_t *stats_capability,
+        _In_    const std::string& stat_enum_str,
+        _In_    const std::string& stat_modes_str)
```
deserialization funciton does not follow convention, only 2 parameters string serialized and output paerameter

rpmarvell replied:
Unlike the other deserialize functions, this stats_capability deserialization has stat_enum and stat_mode_enum as separate strings. Hence it was required to be handled this way. This function works fine in production. Please help , having it this way.

kcudnik commented on meta/SaiSerialize.cpp:
```diff
@@ -5440,3 +5500,79 @@ void sai_deserialize_redis_link_event_damping_aied_config(
+    for (uint32_t i = 0; i < num_capabilities; i++)
+    {
+        /* 1. Populate stat_enum */
+        size_t old_stat_enum_position = stat_enum_position;
+        stat_enum_position = stat_enum_str.find(",", old_stat_enum_position);
+        std::string stat_enum = stat_enum_str.substr(old_stat_enum_position,
+                stat_enum_position - old_stat_enum_position);
+        stats_capability->list[i].stat_enum = std::stoi(stat_enum);
+
```
serialized data should be a json array, and you should easly deserialize that each string or number, there is no necesasry for find string ! pease check list serialize and deserialzie in this file

rpmarvell replied:
Please note that there are some of the existing functions/methods such as sai_deserialize_object_id, sai_deserialize_object_meta_key, sai_deserialize_latch_status, etc are using the "find" and this way of deserializing. As this deserialize(capability get) is going to be used from swss only once during the init time and also it works fine in the production code, kindly allow me to prefer this.

kcudnik commented on meta/SaiSerialize.cpp:
```diff
@@ -5440,3 +5500,79 @@ void sai_deserialize_redis_link_event_damping_aied_config(
+        /* We have run out of values to add to our list */
+        if (stat_enum_position == std::string::npos)
+        {
+            if (num_capabilities != i + 1)
+            {
+                SWSS_LOG_WARN("Lesser stat_enums than expected: expected %d, received %d",
+                        num_capabilities, i+1);
+            }
```
this should throw as data is invalid

rpmarvell replied:
Addressed the comment.

kcudnik commented on syncd/Syncd.cpp:
```diff
@@ -580,6 +583,89 @@ sai_status_t Syncd::processObjectTypeGetAvailabilityQuery(
+    }
+    else if (status == SAI_STATUS_BUFFER_OVERFLOW)
+    {
+        entry =
+        {
+            swss::FieldValueTuple("STAT_COUNT", std::to_string(statCapList.count))
+        };
```
join to 1 line

kcudnik reviewed (changes requested):
address comments

rpmarvell commented on tests/MRVLTL7/query_stats_capability.rec:
```diff
@@ -0,0 +1,6 @@
+2023-06-02.20:39:45.430445|a|INIT_VIEW
+2023-06-02.20:39:45.430645|A|SAI_STATUS_SUCCESS
+2023-06-02.20:39:45.630445|q|stats_capability_query|SAI_OBJECT_TYPE_SWITCH:oid:0x21000000000000|OBJECT_TYPE=SAI_OBJECT_TYPE_QUEUE|LIST_SIZE=1
+2023-06-02.20:39:45.630943|Q|stats_capability_query|SAI_STATUS_BUFFER_OVERFLOW|{"count":12,"list":null}
```
I have added in the recent commit

rpmarvell commented on lib/ClientSai.cpp:
```diff
@@ -999,6 +989,109 @@ sai_status_t ClientSai::waitForGetStatsResponse(
+        return SAI_STATUS_INVALID_PARAMETER;
+    }
+
+    if (stats_capability && stats_capability->list)
```
Addressed this in recent commit

rpmarvell commented on vslib/SwitchStateBase.h:
```diff
@@ -294,6 +294,10 @@ namespace saivs
                               _In_ sai_object_type_t object_type,
                               _In_ sai_attr_id_t attr_id,
                              _Inout_ sai_s32_list_t *enum_values_capability);
+           virtual sai_status_t queryStatsCapability(
```
Addressed the comment in the recent commit.

rpmarvell commented on vslib/SwitchMrvlTL7.cpp:
```diff
@@ -0,0 +1,53 @@
+
+using namespace saivs;
+
+SwitchMrvlTL7::SwitchMrvlTL7(
```
Please note that i have added this as It will be helpful for this testing as well as for the features from Marvell in the future.

rpmarvell commented on vslib/SwitchStateBase.cpp:
```diff
@@ -3812,4 +3811,140 @@ sai_status_t SwitchStateBase::queryAttributeCapability(
+
+        stats_capability->count = 38;
+        stats_capability->list[0].stat_enum = SAI_QUEUE_STAT_PACKETS;
+        stats_capability->list[1].stat_enum = SAI_QUEUE_STAT_BYTES;
+        stats_capability->list[2].stat_enum = SAI_QUEUE_STAT_DROPPED_PACKETS;
+        stats_capability->list[3].stat_enum = SAI_QUEUE_STAT_DROPPED_BYTES;
+        stats_capability->list[4].stat_enum = SAI_QUEUE_STAT_GREEN_PACKETS;
+        stats_capability->list[5].stat_enum = SAI_QUEUE_STAT_GREEN_BYTES;
+        stats_capability->list[6].stat_enum = SAI_QUEUE_STAT_GREEN_DROPPED_PACKETS;
+        stats_capability->list[7].stat_enum = SAI_QUEUE_STAT_GREEN_DROPPED_BYTES;
+        stats_capability->list[8].stat_enum = SAI_QUEUE_STAT_YELLOW_PACKETS;
+        stats_capability->list[9].stat_enum = SAI_QUEUE_STAT_YELLOW_BYTES;
+        stats_capability->list[10].stat_enum = SAI_QUEUE_STAT_YELLOW_DROPPED_PACKETS;
+        stats_capability->list[11].stat_enum = SAI_QUEUE_STAT_YELLOW_DROPPED_BYTES;
+        stats_capability->list[12].stat_enum = SAI_QUEUE_STAT_RED_PACKETS;
+        stats_capability->list[13].stat_enum = SAI_QUEUE_STAT_RED_BYTES;
+        stats_capability->list[14].stat_enum = SAI_QUEUE_STAT_RED_DROPPED_PACKETS;
+        stats_capability->list[15].stat_enum = SAI_QUEUE_STAT_RED_DROPPED_BYTES;
+        stats_capability->list[16].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_DROPPED_PACKETS;
+        stats_capability->list[17].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_DROPPED_BYTES;
+        stats_capability->list[18].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_DROPPED_PACKETS;
+        stats_capability->list[19].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_DROPPED_BYTES;
+        stats_capability->list[20].stat_enum = SAI_QUEUE_STAT_RED_WRED_DROPPED_PACKETS;
+        stats_capability->list[21].stat_enum = SAI_QUEUE_STAT_RED_WRED_DROPPED_BYTES;
+        stats_capability->list[22].stat_enum = SAI_QUEUE_STAT_WRED_DROPPED_PACKETS;
+        stats_capability->list[23].stat_enum = SAI_QUEUE_STAT_WRED_DROPPED_BYTES;
+        stats_capability->list[24].stat_enum = SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES;
+        stats_capability->list[25].stat_enum = SAI_QUEUE_STAT_WATERMARK_BYTES;
+        stats_capability->list[26].stat_enum = SAI_QUEUE_STAT_SHARED_CURR_OCCUPANCY_BYTES;
+        stats_capability->list[27].stat_enum = SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES;
+        stats_capability->list[28].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[29].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[30].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[31].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[32].stat_enum = SAI_QUEUE_STAT_RED_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[33].stat_enum = SAI_QUEUE_STAT_RED_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[34].stat_enum = SAI_QUEUE_STAT_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[35].stat_enum = SAI_QUEUE_STAT_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[36].stat_enum = SAI_QUEUE_STAT_CURR_OCCUPANCY_LEVEL;
+        stats_capability->list[37].stat_enum = SAI_QUEUE_STAT_WATERMARK_LEVEL;
```
Stats mode are set at the for loop below

rpmarvell replied:
Added the queue stats capability assignment in a for loop as suggested. Addressed the review comment.

rpmarvell commented on vslib/SwitchStateBase.cpp:
```diff
@@ -3812,4 +3811,140 @@ sai_status_t SwitchStateBase::queryAttributeCapability(
+        stats_capability->list[11].stat_enum = SAI_PORT_STAT_IF_OUT_NON_UCAST_PKTS;
+        stats_capability->list[12].stat_enum = SAI_PORT_STAT_IF_OUT_DISCARDS;
+        stats_capability->list[13].stat_enum = SAI_PORT_STAT_IF_OUT_ERRORS;
+        stats_capability->list[14].stat_enum = SAI_PORT_STAT_IF_OUT_QLEN;
+        stats_capability->list[15].stat_enum = SAI_PORT_STAT_IF_OUT_BROADCAST_PKTS;
+        stats_capability->list[16].stat_enum = SAI_PORT_STAT_IF_OUT_MULTICAST_PKTS;
+        stats_capability->list[17].stat_enum = SAI_PORT_STAT_ETHER_STATS_DROP_EVENTS;
+        stats_capability->list[18].stat_enum = SAI_PORT_STAT_ETHER_STATS_MULTICAST_PKTS;
+        stats_capability->list[19].stat_enum = SAI_PORT_STAT_ETHER_STATS_BROADCAST_PKTS;
+        stats_capability->list[20].stat_enum = SAI_PORT_STAT_ETHER_STATS_UNDERSIZE_PKTS;
+        stats_capability->list[21].stat_enum = SAI_PORT_STAT_ETHER_STATS_FRAGMENTS;
+        stats_capability->list[22].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_64_OCTETS;
+        stats_capability->list[23].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_65_TO_127_OCTETS;
+        stats_capability->list[24].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_128_TO_255_OCTETS;
+        stats_capability->list[25].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_256_TO_511_OCTETS;
+        stats_capability->list[26].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_512_TO_1023_OCTETS;
+        stats_capability->list[27].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_1024_TO_1518_OCTETS;
+        stats_capability->list[28].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_1519_TO_2047_OCTETS;
+        stats_capability->list[29].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_2048_TO_4095_OCTETS;
+        stats_capability->list[30].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_4096_TO_9216_OCTETS;
+        stats_capability->list[31].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_9217_TO_16383_OCTETS;
+        stats_capability->list[32].stat_enum = SAI_PORT_STAT_ETHER_STATS_OVERSIZE_PKTS;
+        stats_capability->list[33].stat_enum = SAI_PORT_STAT_ETHER_RX_OVERSIZE_PKTS;
+        stats_capability->list[34].stat_enum = SAI_PORT_STAT_ETHER_TX_OVERSIZE_PKTS;
+        stats_capability->list[35].stat_enum = SAI_PORT_STAT_ETHER_STATS_JABBERS;
+        stats_capability->list[36].stat_enum = SAI_PORT_STAT_ETHER_STATS_OCTETS;
+        stats_capability->list[37].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS;
+        stats_capability->list[38].stat_enum = SAI_PORT_STAT_ETHER_STATS_COLLISIONS;
+        stats_capability->list[39].stat_enum = SAI_PORT_STAT_ETHER_STATS_CRC_ALIGN_ERRORS;
+        stats_capability->list[40].stat_enum = SAI_PORT_STAT_ETHER_STATS_TX_NO_ERRORS;
+        stats_capability->list[41].stat_enum = SAI_PORT_STAT_ETHER_STATS_RX_NO_ERRORS;
+        stats_capability->list[42].stat_enum = SAI_PORT_STAT_GREEN_WRED_DROPPED_PACKETS;
+        stats_capability->list[43].stat_enum = SAI_PORT_STAT_GREEN_WRED_DROPPED_BYTES;
+        stats_capability->list[44].stat_enum = SAI_PORT_STAT_YELLOW_WRED_DROPPED_PACKETS;
+        stats_capability->list[45].stat_enum = SAI_PORT_STAT_YELLOW_WRED_DROPPED_BYTES;
+        stats_capability->list[46].stat_enum = SAI_PORT_STAT_RED_WRED_DROPPED_PACKETS;
+        stats_capability->list[47].stat_enum = SAI_PORT_STAT_RED_WRED_DROPPED_BYTES;
+        stats_capability->list[48].stat_enum = SAI_PORT_STAT_WRED_DROPPED_PACKETS;
+        stats_capability->list[49].stat_enum = SAI_PORT_STAT_WRED_DROPPED_BYTES;
+        stats_capability->list[50].stat_enum = SAI_PORT_STAT_ECN_MARKED_PACKETS;
```
I have picked only the Ether,If/port in/out, Ether and wred stat_enums from the sai_port_stat_t. Please note that i would prefer this way unless it is mandatory to add all the stats in sai_port_stat_t in the capability list.

rpmarvell commented on vslib/VirtualSwitchSaiInterface.cpp:
```diff
@@ -972,6 +980,131 @@ sai_status_t VirtualSwitchSaiInterface::queryStatsCapability(
+
+        stats_capability->count = 38;
+        stats_capability->list[0].stat_enum = SAI_QUEUE_STAT_PACKETS;
+        stats_capability->list[1].stat_enum = SAI_QUEUE_STAT_BYTES;
+        stats_capability->list[2].stat_enum = SAI_QUEUE_STAT_DROPPED_PACKETS;
+        stats_capability->list[3].stat_enum = SAI_QUEUE_STAT_DROPPED_BYTES;
+        stats_capability->list[4].stat_enum = SAI_QUEUE_STAT_GREEN_PACKETS;
+        stats_capability->list[5].stat_enum = SAI_QUEUE_STAT_GREEN_BYTES;
+        stats_capability->list[6].stat_enum = SAI_QUEUE_STAT_GREEN_DROPPED_PACKETS;
+        stats_capability->list[7].stat_enum = SAI_QUEUE_STAT_GREEN_DROPPED_BYTES;
+        stats_capability->list[8].stat_enum = SAI_QUEUE_STAT_YELLOW_PACKETS;
+        stats_capability->list[9].stat_enum = SAI_QUEUE_STAT_YELLOW_BYTES;
+        stats_capability->list[10].stat_enum = SAI_QUEUE_STAT_YELLOW_DROPPED_PACKETS;
+        stats_capability->list[11].stat_enum = SAI_QUEUE_STAT_YELLOW_DROPPED_BYTES;
+        stats_capability->list[12].stat_enum = SAI_QUEUE_STAT_RED_PACKETS;
+        stats_capability->list[13].stat_enum = SAI_QUEUE_STAT_RED_BYTES;
+        stats_capability->list[14].stat_enum = SAI_QUEUE_STAT_RED_DROPPED_PACKETS;
+        stats_capability->list[15].stat_enum = SAI_QUEUE_STAT_RED_DROPPED_BYTES;
+        stats_capability->list[16].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_DROPPED_PACKETS;
+        stats_capability->list[17].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_DROPPED_BYTES;
+        stats_capability->list[18].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_DROPPED_PACKETS;
+        stats_capability->list[19].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_DROPPED_BYTES;
+        stats_capability->list[20].stat_enum = SAI_QUEUE_STAT_RED_WRED_DROPPED_PACKETS;
+        stats_capability->list[21].stat_enum = SAI_QUEUE_STAT_RED_WRED_DROPPED_BYTES;
+        stats_capability->list[22].stat_enum = SAI_QUEUE_STAT_WRED_DROPPED_PACKETS;
+        stats_capability->list[23].stat_enum = SAI_QUEUE_STAT_WRED_DROPPED_BYTES;
+        stats_capability->list[24].stat_enum = SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES;
+        stats_capability->list[25].stat_enum = SAI_QUEUE_STAT_WATERMARK_BYTES;
+        stats_capability->list[26].stat_enum = SAI_QUEUE_STAT_SHARED_CURR_OCCUPANCY_BYTES;
+        stats_capability->list[27].stat_enum = SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES;
+        stats_capability->list[28].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[29].stat_enum = SAI_QUEUE_STAT_GREEN_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[30].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[31].stat_enum = SAI_QUEUE_STAT_YELLOW_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[32].stat_enum = SAI_QUEUE_STAT_RED_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[33].stat_enum = SAI_QUEUE_STAT_RED_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[34].stat_enum = SAI_QUEUE_STAT_WRED_ECN_MARKED_PACKETS;
+        stats_capability->list[35].stat_enum = SAI_QUEUE_STAT_WRED_ECN_MARKED_BYTES;
+        stats_capability->list[36].stat_enum = SAI_QUEUE_STAT_CURR_OCCUPANCY_LEVEL;
+        stats_capability->list[37].stat_enum = SAI_QUEUE_STAT_WATERMARK_LEVEL;
```
I have done it in a loop as suggested.

rpmarvell commented on vslib/VirtualSwitchSaiInterface.cpp:
```diff
@@ -972,6 +980,131 @@ sai_status_t VirtualSwitchSaiInterface::queryStatsCapability(
+        stats_capability->list[8].stat_enum = SAI_PORT_STAT_IF_IN_VLAN_DISCARDS;
+        stats_capability->list[9].stat_enum = SAI_PORT_STAT_IF_OUT_OCTETS;
+        stats_capability->list[10].stat_enum = SAI_PORT_STAT_IF_OUT_UCAST_PKTS;
+        stats_capability->list[11].stat_enum = SAI_PORT_STAT_IF_OUT_NON_UCAST_PKTS;
+        stats_capability->list[12].stat_enum = SAI_PORT_STAT_IF_OUT_DISCARDS;
+        stats_capability->list[13].stat_enum = SAI_PORT_STAT_IF_OUT_ERRORS;
+        stats_capability->list[14].stat_enum = SAI_PORT_STAT_IF_OUT_QLEN;
+        stats_capability->list[15].stat_enum = SAI_PORT_STAT_IF_OUT_BROADCAST_PKTS;
+        stats_capability->list[16].stat_enum = SAI_PORT_STAT_IF_OUT_MULTICAST_PKTS;
+        stats_capability->list[17].stat_enum = SAI_PORT_STAT_ETHER_STATS_DROP_EVENTS;
+        stats_capability->list[18].stat_enum = SAI_PORT_STAT_ETHER_STATS_MULTICAST_PKTS;
+        stats_capability->list[19].stat_enum = SAI_PORT_STAT_ETHER_STATS_BROADCAST_PKTS;
+        stats_capability->list[20].stat_enum = SAI_PORT_STAT_ETHER_STATS_UNDERSIZE_PKTS;
+        stats_capability->list[21].stat_enum = SAI_PORT_STAT_ETHER_STATS_FRAGMENTS;
+        stats_capability->list[22].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_64_OCTETS;
+        stats_capability->list[23].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_65_TO_127_OCTETS;
+        stats_capability->list[24].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_128_TO_255_OCTETS;
+        stats_capability->list[25].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_256_TO_511_OCTETS;
+        stats_capability->list[26].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_512_TO_1023_OCTETS;
+        stats_capability->list[27].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_1024_TO_1518_OCTETS;
+        stats_capability->list[28].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_1519_TO_2047_OCTETS;
+        stats_capability->list[29].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_2048_TO_4095_OCTETS;
+        stats_capability->list[30].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_4096_TO_9216_OCTETS;
+        stats_capability->list[31].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS_9217_TO_16383_OCTETS;
+        stats_capability->list[32].stat_enum = SAI_PORT_STAT_ETHER_STATS_OVERSIZE_PKTS;
+        stats_capability->list[33].stat_enum = SAI_PORT_STAT_ETHER_RX_OVERSIZE_PKTS;
+        stats_capability->list[34].stat_enum = SAI_PORT_STAT_ETHER_TX_OVERSIZE_PKTS;
+        stats_capability->list[35].stat_enum = SAI_PORT_STAT_ETHER_STATS_JABBERS;
+        stats_capability->list[36].stat_enum = SAI_PORT_STAT_ETHER_STATS_OCTETS;
+        stats_capability->list[37].stat_enum = SAI_PORT_STAT_ETHER_STATS_PKTS;
+        stats_capability->list[38].stat_enum = SAI_PORT_STAT_ETHER_STATS_COLLISIONS;
+        stats_capability->list[39].stat_enum = SAI_PORT_STAT_ETHER_STATS_CRC_ALIGN_ERRORS;
+        stats_capability->list[40].stat_enum = SAI_PORT_STAT_ETHER_STATS_TX_NO_ERRORS;
+        stats_capability->list[41].stat_enum = SAI_PORT_STAT_ETHER_STATS_RX_NO_ERRORS;
+        stats_capability->list[42].stat_enum = SAI_PORT_STAT_GREEN_WRED_DROPPED_PACKETS;
+        stats_capability->list[43].stat_enum = SAI_PORT_STAT_GREEN_WRED_DROPPED_BYTES;
+        stats_capability->list[44].stat_enum = SAI_PORT_STAT_YELLOW_WRED_DROPPED_PACKETS;
+        stats_capability->list[45].stat_enum = SAI_PORT_STAT_YELLOW_WRED_DROPPED_BYTES;
+        stats_capability->list[46].stat_enum = SAI_PORT_STAT_RED_WRED_DROPPED_PACKETS;
+        stats_capability->list[47].stat_enum = SAI_PORT_STAT_RED_WRED_DROPPED_BYTES;
```
I have picked only the Ether,If/port in/out, Ether and wred stat_enums from the sai_port_stat_t. Please note that i would prefer this way unless it is mandatory to add all the stats in sai_port_stat_t in the capability list.

rpmarvell commented:
@kcudnik  : the vs tests are failing due to the following permission issue, could you help resolve this issue.

sh: 1: ./scripts/pahole-version.sh: Permission denied
init/Kconfig:97: syntax error
init/Kconfig:96: invalid statement
make[1]: *** [scripts/kconfig/Makefile:77: allmodconfig] Error 1
make: *** [Makefile:630: allmodconfig] Error 2

kcudnik commented:
/azp run

kperumalbfn commented:
@rpmarvell Could you rebase to master and fix the build issues.

kcudnik commented on meta/Meta.h:
```diff
@@ -459,10 +459,6 @@ namespace saimeta
                     _Out_ uint64_t *counters,
                     _In_ sai_stats_mode_t mode);
 
-            sai_status_t meta_validate_query_stats_capability(
```
Why this is removed ?

rpmarvell replied:
I have explained this for your comment in the definition of the function.

kcudnik commented on vslib/SwitchStateBase.cpp:
```diff
@@ -3864,3 +3863,103 @@ sai_status_t SwitchStateBase::queryAttributeCapability(
+            return SAI_STATUS_BUFFER_OVERFLOW;
+        }
+
+        stats_capability->count = 51;
```
Please use std vector with declared values of that and then assign in loop

rpmarvell replied:
I have picked only the Ether/If/port in/out and wred stat_enums from the sai_port_stat_t. Hence we need to have unique assignment of these statistics. Please note that i would prefer this way unless it is mandatory to add all the stats in sai_port_stat_t in the capability list.

kperumalbfn commented:
@rpmarvell could you check the code coverage pipeline failure?

kperumalbfn commented:
@rpmarvell Could you take care of code coverage by adding more tests for the changes? We could merge all WRED stats related changes for 202411 release.

![image](https://github.com/user-attachments/assets/6e07ea83-0e31-4635-8e13-01560d2e163d)

kperumalbfn commented:
@rpmarvell Could you resolve the conflicts and check the code coverage. We need to merge this PR first before sonic-swss PRs to avoid build breakage.

kperumalbfn commented:
@rpmarvell Could you take care of sanity checks in the above WRED Sonic PRs. 202411 branch pull will happen soon and would like to merge all WRED ECN stats PRs.

kcudnik commented:
please resolve conflicts

kperumalbfn commented:
@kcudnik could you review and approve this PR

kcudnik commented on vslib/VirtualSwitchSaiInterface.cpp:
```diff
@@ -990,6 +992,94 @@ sai_status_t VirtualSwitchSaiInterface::queryStatsCapability(
+        }
+
+        stats_capability->count = 51;
+        stats_capability->list[0].stat_enum = SAI_PORT_STAT_IF_IN_OCTETS;
```
Create static vector with all stars, and then copy values in a loop

rpmarvell replied:
I have picked only the Ether,If/port in/out, and wred stat_enums from the sai_port_stat_t. Please note that i would prefer this way unless it is mandatory to add all the stats in sai_port_stat_t in the capability list.

kcudnik commented on vslib/SwitchConfig.cpp:
```diff
@@ -105,6 +105,8 @@ bool SwitchConfig::parseSwitchType(
                 boost::algorithm::join(vals, "|").c_str());
 
+
+
```
unnecessary modification to the file

kcudnik commented on vslib/VirtualSwitchSaiInterface.cpp:
```diff
@@ -981,6 +981,8 @@ sai_status_t VirtualSwitchSaiInterface::queryStatsCapability(
 {
     SWSS_LOG_ENTER();
 
+    uint32_t i=0;
```
declare that internally in each for loop defininition, declaring it on function beign is old style C

rpmarvell replied:
Addressed the comment.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

rpmarvell commented:
Tests failing with the following error, looks like a test server issue.

2025-01-17T22:46:16.8973660Z ==============================================================================
2025-01-17T22:46:17.3329451Z Generating script.
2025-01-17T22:46:17.3338692Z ========================== Starting Command Output ===========================
2025-01-17T22:46:17.3367521Z [command]/usr/bin/bash --noprofile --norc /__w/_temp/0e9a17a9-17e3-4a96-9747-e38da04b1e3a.sh
2025-01-17T22:46:17.5177955Z rsyslogd: pidfile '/run/rsyslogd.pid' and pid 15467 already exist.
2025-01-17T22:46:17.5178488Z If you want to run multiple instances of rsyslog, you need to specify
2025-01-17T22:46:17.5178847Z different pid files for them (-i option).
2025-01-17T22:46:17.5179418Z rsyslogd: run failed with error -3000 (see rsyslog.h or try https://www.rsyslog.com/e/3000 to learn what that number means)
2025-01-17T22:46:17.5206996Z 
2025-01-17T22:46:17.5277580Z ##[error]Bash exited with code '1'.
2025-01-17T22:46:17.5399247Z ##[section]Finishing: Update rsyslog.conf

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

rpmarvell commented:
@prsunny @kcudnik  : Could you please help resolve this error "rsyslogd: run failed with error -3000"

mssonicbld commented:
/azp run

kcudnik commented:
this is being handled here: https://github.com/sonic-net/sonic-sairedis/pull/1502

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

rpmarvell commented:
> address comments

I have addressed the comments.

mssonicbld commented:
/azp run

kperumalbfn commented:
/azp run

mssonicbld commented:
/azp run

rpmarvell commented:
@kcudnik : Tests are green. Could you please help with pushing/merging the changes.

r12f commented:
hi @rpmarvell , do you mind to help create a manual PR for this cherry pick? 

202412 branch is in: https://github.com/Azure/sonic-sairedis.msft.

r12f commented:
hi @rpmarvell , gently ping on this one again.

r12f commented:
hi @rpmarvell , do you mind to help create a manual PR for this cherry pick?

202412 branch is in: https://github.com/Azure/sonic-sairedis.msft.
